src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,265 @@
+                                                     GssCtxId** contextHandle,
+                                                     uint32_t isNtlm,
+                                                     GssName* targetName,
+                                                     uint32_t reqFlags,
+                                                     uint8_t* inputBytes,
+                                                     uint32_t inputLength,
+                                                     GssBuffer** outBufferHandle,
+                                                     uint32_t* outTokenLength,
+                                                     uint32_t* retFlags)
+{
+    assert(isNtlm == 0 || isNtlm == 1);
+    assert(minorStatus != nullptr);
+    assert(contextHandle != nullptr);
+    assert(outBufferHandle != nullptr);
+    assert(outTokenLength != nullptr);
+    assert(retFlags != nullptr);
+
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID desiredMech = isNtlm ? GSS_NTLM_MECHANISM : GSS_SPNEGO_MECHANISM;
+#else
+    assert(!isNtlm && "NTLM is not supported by MIT libgssapi_krb5");
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .elements = static_cast<void*>(gss_mech_value)};
+    gss_OID desiredMech = &gss_mech_spnego_OID_desc;
+#endif
+
+    gss_buffer_desc inputToken{.length = UnsignedCast(inputLength), .value = inputBytes};
+    *outBufferHandle = new GssBuffer();
+
+    uint32_t majorStatus = gss_init_sec_context(minorStatus,
+                                                claimantCredHandle,
+                                                contextHandle,
+                                                targetName,
+                                                desiredMech,
+                                                reqFlags,
+                                                0,
+                                                GSS_C_NO_CHANNEL_BINDINGS,
+                                                &inputToken,
+                                                nullptr,
+                                                *outBufferHandle,
+                                                retFlags,
+                                                nullptr);
+
+    NetSecurityNative_HandleError(majorStatus, outBufferHandle, outTokenLength);
+    return majorStatus;
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseCred(uint32_t* minorStatus, GssCredId** credHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(credHandle != nullptr);
+    return gss_release_cred(minorStatus, credHandle);
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseBuffer(uint32_t* minorStatus, GssBuffer* buffer)
+{
+    assert(minorStatus != nullptr);
+    assert(buffer != nullptr);
+    uint32_t status = gss_release_buffer(minorStatus, buffer);
+    delete buffer;
+    return status;
+}
+
+extern "C" void
+NetSecurityNative_CopyBuffer(GssBuffer* bufferHandle, uint8_t* bytes, uint32_t capacity, uint32_t offset)
```

Comment:
We typically order such arguments as offset and then length.  It'd also be good to choose names that'll help the caller understand what's what, e.g. ```sourceBufferHandle``` and ```destination``` instead of ```bufferHandle``` and ```bytes```.